compiler/rustc_lint/src/redundant_lifetime_args.rs

```diff
@@ -0,0 +1,177 @@
+use rustc_infer::infer::{SubregionOrigin, TyCtxtInferExt};
+use rustc_macros::LintDiagnostic;
+use rustc_middle::ty::{self, TyCtxt};
+use rustc_session::lint::builtin::UNUSED_LIFETIMES;
+use rustc_span::DUMMY_SP;
+use rustc_trait_selection::traits::{outlives_bounds::InferCtxtExt, ObligationCtxt};
+
+use crate::{LateContext, LateLintPass};
+
+declare_lint_pass!(RedundantLifetimeArgs => []);
+
+impl<'tcx> LateLintPass<'tcx> for RedundantLifetimeArgs {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx hir::Item<'tcx>) {
+        check(cx.tcx, cx.param_env, item.owner_id);
+    }
+
+    fn check_trait_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx hir::TraitItem<'tcx>) {
+        check(cx.tcx, cx.param_env, item.owner_id);
+    }
+
+    fn check_impl_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx hir::ImplItem<'tcx>) {
+        if cx
+            .tcx
+            .hir()
+            .expect_item(cx.tcx.local_parent(item.owner_id.def_id))
+            .expect_impl()
+            .of_trait
+            .is_some()
+        {
+            // Don't check for redundant lifetimes for trait implementations,
+            // since the signature is required to be compatible with the trait.
+            return;
+        }
+
+        check(cx.tcx, cx.param_env, item.owner_id);
+    }
+}
+
+fn check<'tcx>(tcx: TyCtxt<'tcx>, param_env: ty::ParamEnv<'tcx>, owner_id: hir::OwnerId) {
+    let def_kind = tcx.def_kind(owner_id);
+    match def_kind {
+        DefKind::Struct
+        | DefKind::Union
+        | DefKind::Enum
+        | DefKind::Trait
+        | DefKind::TraitAlias
+        | DefKind::AssocTy
+        | DefKind::Fn
+        | DefKind::Const
+        | DefKind::AssocFn
+        | DefKind::AssocConst
+        | DefKind::Impl { of_trait: _ } => {
+            // Proceed
+        }
+        DefKind::Mod
+        | DefKind::Variant
+        | DefKind::TyAlias
+        | DefKind::ForeignTy
+        | DefKind::TyParam
+        | DefKind::ConstParam
+        | DefKind::Static(_)
+        | DefKind::Ctor(_, _)
+        | DefKind::Macro(_)
+        | DefKind::ExternCrate
+        | DefKind::Use
+        | DefKind::ForeignMod
+        | DefKind::AnonConst
+        | DefKind::InlineConst
+        | DefKind::OpaqueTy
+        | DefKind::Field
+        | DefKind::LifetimeParam
+        | DefKind::GlobalAsm
+        | DefKind::Closure => return,
+    }
+
+    let infcx = &tcx.infer_ctxt().build();
+    let ocx = ObligationCtxt::new(infcx);
+
+    // Compute the implied outlives bounds for the item. This ensures that we treat
+    // a signature with an argument like `&'a &'b ()` as implicitly having `'b: 'a`.
+    let Ok(assumed_wf_types) = ocx.assumed_wf_types(param_env, owner_id.def_id) else {
+        return;
+    };
+    let implied_bounds = infcx.implied_bounds_tys(param_env, owner_id.def_id, assumed_wf_types);
+    let outlives_env = &OutlivesEnvironment::with_bounds(param_env, implied_bounds);
+
```

Comment:
I have implemented a similar thing in https://github.com/rust-lang/rust/pull/116891/commits/350b33ee7aabbea131a63217f4f610bcde81cfe7. The algorithm is different though. I have customized it for this case. Feel free to use it if you like.

```suggestion
    let mut candidates = vec![tcx.lifetimes.re_static];
    
    for lifetime_arg in Iterator::chain(
        ty::GenericArgs::identity_for_item(tcx, owner_id).iter().filter_map(|arg| arg.as_region()),
        // If we are in a function, consider its late-bound lifetimes too.
        matches!(def_kind, DefKind::Fn | DefKind::AssocFn)
            .then(|| tcx.fn_sig(owner_id).instantiate_identity().bound_vars())
            .into_iter()
            .flatten()
            .filter_map(|var| {
                let ty::BoundVariableKind::Region(kind) = var else { return None };
                Some(ty::Region::new_late_param(tcx, owner_id.to_def_id(), kind))
            }),
    ) {
        if !lifetime_arg.has_name() {
            continue;
        }
    
        let lifetime_def_id = match lifetime_arg.kind() {
            ty::ReEarlyParam(ty::EarlyParamRegion { def_id, .. }) => def_id,
            ty::ReLateParam(late) => late.bound_region.get_id().unwrap(),
            _ => bug!("not a lifetime parameter {lifetime_arg:?}"),
        };
    
        if tcx.parent(lifetime_def_id) = owner_id.to_def_id() {
            // Do not rename lifetimes not local to this item since they'll overlap
            // with the lint running on the parent. We still want to consider parent
            // lifetimes which make child lifetimes redundant, otherwise we would
            // have truncated the `identity_for_item` args above.
        } else if let Some(&candidate) = candidates.iter().find(|&&candidate| {
            let free_regions = outlives_env.free_region_map();
            free_regions.sub_free_regions(tcx, candidate, lifetime_arg)
                && free_regions.sub_free_regions(tcx, lifetime_arg, candidate)
        }) {
            tcx.emit_spanned_lint(
                UNUSED_LIFETIMES,
                tcx.local_def_id_to_hir_id(lifetime_def_id.expect_local()),
                tcx.def_span(lifetime_def_id),
                RedundantLifetimeArgsLint { candidate, victim: lifetime_arg },
            );
        }
    
        candidates.push(lifetime_arg);
    }
```